File: projects/m5/002-polygon-area-calculator/solutions/enricozoboli/shape_calculator.py
from __future__ import annotations


class Rectangle:
    def __init__(self, width: int | float, height: int | float) -> None:
        self.width = width
        self.height = height

    def __str__(self) -> str:
        return f"{self.__class__.__name__}(width={self.width},"\
               f" height={self.height})"

    def set_width(self, new_width: int | float) -> None:
        self.width = new_width

    def set_height(self, new_height: int | float) -> None:
        self.height = new_height

    def get_area(self) -> int | float:
        return self.width * self.height

    def get_perimeter(self) -> int | float:
        return 2 * self.width + 2 * self.height

    def get_diagonal(self) -> int | float:
        return (self.width ** 2 + self.height ** 2) ** 0.5

    def get_picture(self) -> str:
        output_pic = ""
        # without int casting the typechecker throw an incompatibility error.
        line_range = int(self.height)
        if self.width <= 50 and self.height <= 50:
            for line in range(line_range):
                output_pic += '*' * int(self.width) + '\n'
        else:
            output_pic += "Too big for picture."
        return output_pic

    def get_amount_inside(self, shape: Rectangle) -> int:
        height_fit = self.height // shape.height
        width_fit = self.width // shape.width
        return int(height_fit) * int(width_fit)


class Square(Rectangle):
    def __init__(self, side: int | float) -> None:
        super().__init__(side, side)
        self.side = side

    def __str__(self) -> str:
        return f"{self.__class__.__name__}(side={self.side})"

    def set_side(self, new_side: int | float) -> None:
        self.side = new_side
        Rectangle.set_width(self, new_side)
        Rectangle.set_height(self, new_side)

    def set_width(self, new_width: int | float) -> None:
        self.side = new_width

    def set_height(self, new_height: int | float) -> None:
        self.side = new_height

File: projects/m5/002-polygon-area-calculator/solutions/LucaBertoldi/shape_calculator.php
<?php

class Rectangle {
    public function __construct(int $width, int $height) {
        $this->width = $width;
        $this->height = $height;
    }

    public function __toString() {
        return get_class($this) . "(width=$this->width, height=$this->height)";
    }

    public function set_width(int $width) {
        return $this->width = $width;
    }

    public function set_height(int $height) {
        return $this->height = $height;
    }

    public function get_area() {
        
        return $this->width * $this->height;
    }

    public function get_perimeter() {
        return ($this->width * $this->height)*2;
    }

    public function get_diagonal() {
        return sqrt(pow($this->width, 2) + pow($this->height, 2));
    }

    public function get_picture () {
        $line = "";
        for($i = 0; $i < $this->height; $i++) {
            for ($j = 0; $j <= $this->width; $j++) {
                $line .= "*";
            }
            $line .= "<br/>";
        }
         
        return $line;
    }

    public function get_amount_inside ($shape) {
        return $this->get_area() / $shape->get_area();
    }

    
};


class Square extends Rectangle {
    public function __construct(int $side) {
        $this->side = $side;
        $this->_init_();
    }

    protected function _init_() {
        $this->width = $this->side;
        $this->height = $this->side;
    }

    public function set_side(int $side) {
        $this->width = $this->set_width($side);
    }

    public function set_width(int $width) {
        $this->width = $width;
        $this->height = $width;
    }

    public function set_height(int $height) {
        $this->width = $height;
        $this->height = $height;
    }
}

?>




